<a href="https://www.kaggle.com/code/houmannorasteh/mpqa-on-srl?scriptVersionId=175122373" target="_blank"><img align="left" alt="Kaggle" title="Open in Kaggle" src="https://kaggle.com/static/images/open-in-kaggle.svg"></a>

In [1]:
! pip install transformers
! pip install torch

### importing libraries 

In [2]:
import torch
import numpy as np
import pandas as pd
from transformers import T5Tokenizer, T5ForConditionalGeneration
import json
from itertools import zip_longest

### defining general variables

In [3]:
#defining global valriables throughout the whole notebook
EPOCH = 10
BATCH_SIZE = 16
MAX_INPUT_LENGTH = 65
MAX_LABEL_LENGTH = 8
MODEL_LINK = "google/flan-t5-small"
FOLD_NUMBER = 4

In [4]:
tokenizer = T5Tokenizer.from_pretrained(MODEL_LINK, legacy=False)
model = T5ForConditionalGeneration.from_pretrained(MODEL_LINK).to('cuda')

Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.


### functions created for reading(get_data) and organize the files (organize_data)

In [5]:
# a funciton to read data off of a database link is here to help getting and organizing data into dataframes
def get_data(address):
    lines = []
    with open(address) as file:
        for line in file:
            x = json.loads(line)
            lines.append(x)
    sentences, orl, sep_sentences = [], [], []
    for i in range(len(lines)):
        sep_sentences.append(lines[i]['sentences'])
        sentences.append(' '.join(lines[i]['sentences']))
        orl.append(lines[i]['orl'])
    dataframe = pd.DataFrame({'sentence': sentences, 'orl': orl, 'sep_sent': sep_sentences})
    return dataframe

# this function is to make a list of the said attribute for later iterations
def list_of(attributes, requested_atr):
    requested_list = []
    for sublist in attributes:
        if sublist[-1] == requested_atr:
            requested_list.append(sublist)
    return requested_list

# this function was made to find target(s)/agent(s) of a dse according to list of attributes
def organize_data(attributes, sentence):
    AGENT, DSE, TARGET = '', '', ''
    target_flag, agent_flag = False, False
    for sublist in attributes:
        if sublist[-1] == 'DSE':
            dse_start = int(sublist[0])
            dse_end = int(sublist[1] + 1)
            DSE += ' '.join(sentence[dse_start:dse_end]) + '|'
            
            
            # looking for the targets and agents of this dse that we have found
            target_flag = False
            for sub_sublist in list_of(attributes, 'TARGET'):
                if sub_sublist[0] == dse_start and int(sub_sublist[1] + 1) == dse_end:
                    target_start = int(sub_sublist[2])
                    target_end = int(sub_sublist[3] + 1)
                    TARGET += ' '.join(sentence[target_start:target_end]) + '|'
                    target_flag = True
            if not target_flag:
                TARGET += ' |'
            
            agent_flag = False
            for sub_sublist in list_of(attributes, 'AGENT'):
                if sub_sublist[0] == dse_start and int(sub_sublist[1] + 1) == dse_end:
                    agent_start = int(sub_sublist[2])
                    agent_end = int(sub_sublist[3] + 1)
                    AGENT += ' '.join(sentence[agent_start:agent_end]) + '|'
                    agent_flag = True
            if not agent_flag:
                AGENT += ' |'
    return AGENT, DSE, TARGET

### a function to call for different links of folders with exact process to extract and organize data

In [6]:
def get_files_of_folder(folder_number):
    folder_number = int()
    folder = f"/kaggle/input/ds-json-format/json_format_dataset/{folder_number}"
    dev_df = get_data(f"{folder}/aaai19srl.dev{folder_number}.conll.json")
    df = get_data(f"{folder}/aaai19srl.train{folder_number}.conll.json")
    test_df = get_data(f"{folder}/aaai19srl.test{folder_number}.conll.json")
    return df, dev_df, test_df
df = get_files_of_folder(FOLD_NUMBER)[0]
dev_df = get_files_of_folder(FOLD_NUMBER)[1]
test_df = get_files_of_folder(FOLD_NUMBER)[2]

### getting training/validation/test data into data frames and dividing each of {agent, target, dse} elements

In [7]:
# df = get_data("/kaggle/input/ds-json-format/json_format_dataset/0/aaai19srl.train0.conll.json")

for i in range(len(df)):
    agent, dse, target = organize_data(df['orl'][i], df['sep_sent'][i])
    df.loc[i, 'agent'] = agent
    df.loc[i, 'dse'] = dse
    df.loc[i, 'target'] = target

df.drop(columns='sep_sent')

sentence  \
0     The Kimberley Provincial Hospital said it woul...   
1     Saeed said indications were that those tests w...   
2     He said it was his opinion that the patient --...   
3     The woman was admitted to the hospital on Satu...   
4     `` Since our technical equipment is far from p...   
...                                                 ...   
2444  Benjamin Franklin Federal Savings & Loan Assoc...   
2445  thrift said the restructuring should help it m...   
2446  Details of the restructuring wo n't be made fi...   
2447  Jay Stevens , an analyst with Dean Witter Reyn...   
2448  company to earn 35 cents a share for the quart...   

                                                    orl  \
0     [[6, 8, 0, 3, AGENT], [6, 8, 6, 8, DSE], [6, 8...   
1     [[1, 1, 0, 0, AGENT], [1, 1, 1, 1, DSE], [1, 1...   
2     [[4, 5, 0, 0, AGENT], [4, 5, 4, 5, DSE], [4, 5...   
3     [[10, 10, 0, 1, AGENT], [10, 10, 10, 10, DSE],...   
4     [[22, 22, 2, 4, TARGET], [22, 22, 10, 10, TARG...   
...                                                 ...   
2444           [[9, 9, 8, 8, AGENT], [9, 9, 9, 9, DSE]]   
2445          [[1, 1, 1, 1, DSE], [1, 1, 2, 3, TARGET]]   
2446   [[11, 11, 10, 10, AGENT], [11, 11, 11, 11, DSE]]   
2447   [[12, 12, 11, 11, AGENT], [12, 12, 12, 12, DSE]]   
2448  [[11, 11, 11, 11, DSE], [11, 11, 12, 16, TARGET]]   

                                   agent                   dse  \
0     The Kimberley Provincial Hospital|  would probably know|   
1                                 Saeed|                 said|   
2                                    He|          his opinion|   
3                             The woman|          complaining|   
4                               Nazarov|                 said|   
...                                  ...                   ...   
2444                                 it|                plans|   
2445                                   |                 said|   
2446                         regulators|              approve|   
2447                                 he|             expected|   
2448                                   |                 said|   

                                            target  
0     whether one of its patients had Congo Fever|  
1                                     those tests|  
2                          the patient -- a woman|  
3                              severe joint pains|  
4                      our technical equipment|we|  
...                                            ...  
2444                                             |  
2445                            the restructuring|  
2446                                             |  
2447                                             |  
2448                    the firm 's weaker profit|  

[2449 rows x 5 columns]

In [8]:
# dev_df = get_data("/kaggle/input/ds-json-format/json_format_dataset/0/aaai19srl.dev0.conll.json")

for i in range(len(dev_df)):
    agent, dse, target = organize_data(dev_df['orl'][i], dev_df['sep_sent'][i])
    dev_df.loc[i, 'agent'] = agent
    dev_df.loc[i, 'dse'] = dse
    dev_df.loc[i, 'target'] = target

dev_df.drop(columns='sep_sent')

sentence  \
0     The owner though that the animal was suffering...   
1     The owner put down the animal , although the v...   
2     GATUNA , Rwanda , July 6 -LRB- AFP -RRB- - Pre...   
3     The formerly close allies fell out in 1999 , t...   
4     In March , Uganda declared Rwanda a hostile na...   
...                                                 ...   
1033  He said the independent power segment could gr...   
1034  Westinghouse also expects its international sa...   
1035  Analysts had expected Consolidated to post a s...   
1036  They have to continue to tighten their belts ,...   
1037  The move rewards shareholders and should impro...   

                                                    orl  \
0     [[2, 2, 0, 1, AGENT], [2, 2, 2, 2, DSE], [2, 2...   
1     [[10, 11, 8, 9, AGENT], [10, 11, 10, 11, DSE],...   
2     [[30, 32, 30, 32, DSE], [30, 32, 33, 35, TARGET]]   
3     [[4, 5, 0, 3, AGENT], [4, 5, 4, 5, DSE], [4, 5...   
4     [[4, 4, 3, 3, AGENT], [4, 4, 4, 4, DSE], [4, 4...   
...                                                 ...   
1033  [[1, 1, 0, 0, AGENT], [1, 1, 1, 1, DSE], [1, 1...   
1034           [[2, 2, 0, 0, AGENT], [2, 2, 2, 2, DSE]]   
1035  [[2, 2, 0, 0, AGENT], [2, 2, 2, 2, DSE], [2, 2...   
1036  [[10, 10, 0, 0, TARGET], [10, 10, 10, 10, DSE]...   
1037  [[13, 13, 0, 1, TARGET], [13, 13, 12, 12, AGEN...   

                                                  agent  \
0                                            The owner|   
1                                              the vet|   
2                                                     |   
3     The formerly close allies|The formerly close a...   
4                                           Uganda| | |   
...                                                 ...   
1033                                             He|We|   
1034                                      Westinghouse|   
1035                                          Analysts|   
1036  Craig Kloner , an analyst at Goldman , Sachs &...   
1037                                            Oneida|   

                             dse  \
0                        though|   
1                 had forbidden|   
2      soured relations between|   
3     fell out|mounting rivalry|   
4      declared|alleged|support|   
...                          ...   
1033                said|expect|   
1034                    expects|   
1035                   expected|   
1036                       said|   
1037                       said|   

                                                 target  
0                                           the animal|  
1                                         him to do so|  
2                         their neighbouring countries|  
3                                               each| |  
4     Rwanda|Kigali|a rival to Museveni in a preside...  
...                                                 ...  
1033  the independent power segment|to supply a sign...  
1034                                                  |  
1035                Consolidated to post a slim profit|  
1036                                              They|  
1037                                          The move|  

[1038 rows x 5 columns]

In [9]:
# test_df = get_data("/kaggle/input/ds-json-format/json_format_dataset/0/aaai19srl.test0.conll.json")

for i in range(len(test_df)):
    agent, dse, target = organize_data(test_df['orl'][i], test_df['sep_sent'][i])
    test_df.loc[i, 'agent'] = agent
    test_df.loc[i, 'dse'] = dse
    test_df.loc[i, 'target'] = target

test_df.drop(columns='sep_sent')

sentence  \
0    Private organizations are also being encourage...   
1    It is believed that sand from northern areas ,...   
2    Russia favors creation of `` international ins...   
3    Moscow , 16 July : The Russian Foreign Ministr...   
4    A Foreign Ministry statement in connection wit...   
..                                                 ...   
620  We will have a reasonably flat year this year ...   
621  Mr. Schwarz also said costs associated with U....   
622  Esselte AB , the Stockholm office supplies com...   
623  A committee of outside directors for the Garde...   
624  The unit said it can provide no assurance a tr...   

                                                   orl  \
0            [[2, 5, 2, 5, DSE], [2, 5, 7, 9, TARGET]]   
1           [[1, 2, 1, 2, DSE], [1, 2, 4, 25, TARGET]]   
2    [[1, 1, 0, 0, AGENT], [1, 1, 1, 1, DSE], [1, 1...   
3    [[9, 11, 5, 8, AGENT], [9, 11, 9, 11, DSE], [9...   
4    [[25, 25, 0, 3, AGENT], [25, 25, 25, 25, DSE],...   
..                                                 ...   
620  [[13, 13, 4, 6, TARGET], [13, 13, 11, 12, AGEN...   
621  [[3, 3, 0, 1, AGENT], [3, 3, 3, 3, DSE], [3, 3...   
622  [[12, 12, 0, 8, AGENT], [12, 12, 12, 12, DSE],...   
623   [[20, 20, 18, 19, AGENT], [20, 20, 20, 20, DSE]]   
624  [[4, 7, 0, 1, AGENT], [4, 7, 4, 7, DSE], [4, 7...   

                                                 agent  \
0                                                    |   
1                                                    |   
2                                              Russia|   
3                        The Russian Foreign Ministry|   
4                        A Foreign Ministry statement|   
..                                                 ...   
620                                       Mr. Schwarz|   
621                                       Mr. Schwarz|   
622  Esselte AB , the Stockholm office supplies com...   
623                                        the parent|   
624                                          The unit|   

                            dse  \
0    are also being encouraged|   
1                  is believed|   
2                       favors|   
3               has called for|   
4                        reads|   
..                          ...   
620                       says|   
621                       said|   
622                   proposed|   
623                      asked|   
624   can provide no assurance|   

                                                target  
0                               help fight sandstorms|  
1    sand from northern areas , especially the Inne...  
2    creation of `` international instruments '' to...  
3    taking into account Russia 's specific interes...  
4             taking into account Russia 's interests|  
..                                                 ...  
620                              reasonably flat year|  
621  costs associated with U.S. Trust 's planned move|  
622  to acquire the 22 % it does n't own of its U.S...  
623                                                  |  
624                          a transaction will occur|  

[625 rows x 5 columns]

### Getting evaluation data into df and dividing each of {agent, target, dse} elements

In [10]:
def pipeDivider(pipedString):
    listOfItems = []
    listOfItems = pipedString.split('|')[:-1]
    return listOfItems

In [11]:
# function which will write prompt for the model according to the sentence and the items in it
def create_prompt(input_df):
    output_df = pd.DataFrame(columns=['target_prompt', 'agent_prompt', 'target', 'agent'])
    for i in range(len(input_df)):
        dse_list, target_list, agent_list = [], [], []
        dse_list = pipeDivider(str(input_df.iloc[i]['dse']))
        target_list = pipeDivider(str(input_df.iloc[i]['target']))
        agent_list = pipeDivider(str(input_df.iloc[i]['agent']))
        for j in range(len(dse_list)):
            last_row = int(len(output_df))+1
            output_df.loc[last_row, 'target_prompt'] = f"Sentence is: {input_df.iloc[i]['sentence']} Find target for this dse: {dse_list[j]}"
            output_df.loc[last_row, 'agent_prompt'] = f"Sentence is: {input_df.iloc[i]['sentence']} Find agent for this dse: {dse_list[j]}"
            output_df.loc[last_row, 'target'] = target_list[j]
            output_df.loc[last_row, 'agent'] = agent_list[j]
    return output_df

In [12]:
train_df = create_prompt(df)
train_df.head(-10)

target_prompt  \
1     Sentence is: The Kimberley Provincial Hospital...   
2     Sentence is: Saeed said indications were that ...   
3     Sentence is: He said it was his opinion that t...   
4     Sentence is: The woman was admitted to the hos...   
5     Sentence is: `` Since our technical equipment ...   
...                                                 ...   
3521  Sentence is: The transaction has been approved...   
3522  Sentence is: The transaction has been approved...   
3523  Sentence is: holders adopted a shareholder-rig...   
3524  Sentence is: Earlier this month , the company ...   
3525  Sentence is: McNealy said the issues that hurt...   

                                           agent_prompt  \
1     Sentence is: The Kimberley Provincial Hospital...   
2     Sentence is: Saeed said indications were that ...   
3     Sentence is: He said it was his opinion that t...   
4     Sentence is: The woman was admitted to the hos...   
5     Sentence is: `` Since our technical equipment ...   
...                                                 ...   
3521  Sentence is: The transaction has been approved...   
3522  Sentence is: The transaction has been approved...   
3523  Sentence is: holders adopted a shareholder-rig...   
3524  Sentence is: Earlier this month , the company ...   
3525  Sentence is: McNealy said the issues that hurt...   

                                           target  \
1     whether one of its patients had Congo Fever   
2                                     those tests   
3                          the patient -- a woman   
4                              severe joint pains   
5                         our technical equipment   
...                                           ...   
3521                              The transaction   
3522                                                
3523                    a shareholder-rights plan   
3524                to break even for the quarter   
3525                         Sun 's profitability   

                                  agent  
1     The Kimberley Provincial Hospital  
2                                 Saeed  
3                                    He  
4                             The woman  
5                               Nazarov  
...                                 ...  
3521                      Kyle 's board  
3522        the company 's shareholders  
3523                                     
3524                                 it  
3525                                 he  

[3525 rows x 4 columns]

### Change the sentence and dse into a prompt according to information in that row 
#### change will be applied on all three data frames {train, test, validation}

In [13]:
dev_df = create_prompt(dev_df)
train_df = create_prompt(df)
test_df = create_prompt(test_df)

In [14]:
train_df = train_df.reset_index(drop=True)
dev_df = dev_df.reset_index(drop=True)
test_df = test_df.reset_index(drop=True)

In [15]:
from torch.utils.data import DataLoader, Dataset
# Define a custom dataset
class CustomDataset(Dataset):
    def __init__(self, dataframe, tokenizer, max_length4text, max_length4label):
        self.data = dataframe
        self.tokenizer = tokenizer
        self.max_length4text = max_length4text
        self.max_length4label = max_length4label

    def __len__(self):
        return len(self.data)

    def __getitem__(self, idx):
        agent_prompt = self.data['agent_prompt'][idx]
        target_prompt = self.data['target_prompt'][idx]
        agent = self.data['agent'][idx]
        target = self.data['target'][idx]
        # tokenizing agent prompt
        agent_prompt_encoding = self.tokenizer(agent_prompt, truncation=True, padding='max_length', max_length=self.max_length4text, return_tensors='pt').to("cuda")
        # tokenizing target prompt
        target_prompt_encoding = self.tokenizer(target_prompt, truncation=True, padding='max_length', max_length=self.max_length4text, return_tensors='pt').to("cuda")
        # tokenizing agent
        agent_encoding = self.tokenizer(agent, truncation=True, padding='max_length', max_length=self.max_length4label, return_tensors='pt').to("cuda")
        # tokenizing target
        target_encoding = self.tokenizer(target, truncation=True, padding='max_length', max_length=self.max_length4label, return_tensors='pt').to("cuda")
        return {
            # agent prompt
            'agent_input_id': agent_prompt_encoding['input_ids'].squeeze(),
            'agent_attention_mask': agent_prompt_encoding['attention_mask'].squeeze(),
            # target prompt
            'target_input_id': target_prompt_encoding['input_ids'].squeeze(),
            'target_attention_mask': target_prompt_encoding['attention_mask'].squeeze(),
            #agent
            'agent_id': agent_encoding['input_ids'].squeeze(),
            'agent_mask': agent_encoding['attention_mask'].squeeze(),
            #target
            'target_id': target_encoding['input_ids'].squeeze(),
            'target_mask': target_encoding['attention_mask'].squeeze()
        }

In [25]:
# Create a data loader for TRAIN dataframe 
train_dataset = CustomDataset(train_df[:500], tokenizer, max_length4text= MAX_INPUT_LENGTH, max_length4label= MAX_LABEL_LENGTH)
train_data_loader = DataLoader(train_dataset, batch_size= BATCH_SIZE, shuffle=True)

# Create a data loader for EVALUATION dataframe
val_dataset = CustomDataset(dev_df, tokenizer, max_length4text= MAX_INPUT_LENGTH, max_length4label= MAX_LABEL_LENGTH)
val_data_loader = DataLoader(val_dataset, batch_size= BATCH_SIZE, shuffle=False)

# Create a data loader for TEST dataframe
test_dataset = CustomDataset(test_df, tokenizer, max_length4text= MAX_INPUT_LENGTH, max_length4label= MAX_LABEL_LENGTH)
test_data_loader = DataLoader(test_dataset, batch_size= BATCH_SIZE, shuffle=False)



In [17]:
# Define the loss function and optimizer
loss_fn = torch.nn.CrossEntropyLoss()
optimizer = torch.optim.Adam(model.parameters(), lr=0.001)

In [18]:
def f1_calculator(pred_list, actual_list):
    cleaned_pred, cleaned_actual = [], []
    matched = 0
    for i in range(len(pred_list)):
        for j in range(len(actual_list)):
            if actual_list[j] in pred_list:
                matched += 1
        cleaned_pred.extend(pred_list)
        cleaned_actual.extend(actual_list)
    
    prediction_len = len(cleaned_pred)
    actual_len = len(cleaned_actual)
    print(f'matched: {matched}, prediction_len:{prediction_len}, actual_len:{actual_len} \n')
    try:
        precision = (matched / prediction_len)
        recall = (matched / actual_len)
        f1 = (2 * (precision * recall)) / (precision + recall)
    except:
        f1 = 0
    return f1

In [27]:
mpqa_trained_output = pd.DataFrame()

ValueError: columns cannot be a set

In [34]:
mpqa_trained_output = pd.DataFrame(['mpqa', 'mpqa_model_output'])
def evaluate_model(model, dataloader, prompt_type):
    if prompt_type == 'target':
        id_type = 'target_input_id'
        attention_type = 'target_attention_mask'
        output_type = 'target_id'
    elif prompt_type == 'agent':
        id_type = 'agent_input_id'
        attention_type = 'agent_attention_mask'
        output_type = 'agent_id'
    
    model.eval()  # Set the model to evaluation mode
    with torch.no_grad():
        true_counted, prediction_counted, actual_counted = 0,0,0
        matched, prediction_len, actual_len = 0,0,0
        actual_list, prediction_list, f1 = [], [], []
        for batch_idx, batch in enumerate(dataloader):
            # Move data to the specified device
            # batch = {key: value.to('cuda') for key, value in batch.items()}

            # Forward pass
            ids = batch[id_type]
            mask = batch[attention_type]
            output_id = batch[output_type]
            
            actuals = [tokenizer.decode(g, skip_special_tokens=True, clean_up_tokenization_spaces=True) for g in output_id]
            
            generated_output = generated_ids = model.generate(
              input_ids = ids,
              attention_mask = mask, 
              max_length=64, 
              )
            preds = [tokenizer.decode(g, skip_special_tokens=True, clean_up_tokenization_spaces=True) for g in generated_output]
            
            actual_list.extend(actuals)
            prediction_list.extend(preds)
            temprorarly_pd = pd.DataFrame(columns={'mpqa':actual_list, 'mpqa_model_output':prediction_list})
        mpqa_trained_output = mpqa_trained_output.append(temprorarly_pd)
    return f1_calculator(prediction_list, actual_list)

In [35]:
def train_data(data_loader, v_data_loader):
    model.train()
    for epoch in range(EPOCH):
        losses = []
        for batch in data_loader:

            agent_input = batch['agent_input_id']
            agent_attention_mask = batch['agent_attention_mask']

            target_input = batch['target_input_id']
            target_attention_mask = batch['target_attention_mask']

            agent_id = batch['agent_id']
            agent_mask = batch['agent_mask']

            target_id = batch['target_id']
            target_mask = batch['target_mask']

            optimizer.zero_grad()

            agent_output = model(agent_input, attention_mask=agent_attention_mask, labels=agent_id)
            target_output = model(target_input, attention_mask=target_attention_mask, labels=target_id)

            agent_loss = agent_output.loss
            target_loss = target_output.loss
            losses.append(agent_loss.item())
            losses.append(target_loss.item())

            agent_loss.backward()
            target_loss.backward()
            optimizer.step()

        f1_4_target = evaluate_model(model, v_data_loader, 'target')
        f1_4_agent = evaluate_model(model, v_data_loader, 'agent')
        print(f'loss: {np.mean(losses)}, f1 for target:{f1_4_target}, f1 for agent:{f1_4_agent} \n end of epoch{epoch}. \n')

### Applying 10 percent of SRL to the trained data to see how good or bad does it perform

##### importing and cleaning data for training ( skip if you have model )

In [30]:
srl = pd.read_csv('/kaggle/input/10-of-srl/output.csv')
agents, targets, verbs, sentences = [], [], [], []
for i in range(len(srl)):
    line = srl.iloc[i]
    if len(line['arguments']) > 2:
        print(f'line:{i} is {line['arguments']}')
        temp = line['arguments'].split('|')
        agents.append(temp[0])
        targets.append(temp[1])
    else:
        agents.append('')
        targets.append('')
    sentences.append(line['sentence'])
    verbs.append(line['predicate'])

srl_df = pd.DataFrame({'sentences': sentences, 'verbs': verbs, 'agents': agents, 'targets': targets})
srl_df.head()

SyntaxError: f-string: unmatched '[' (2699868172.py, line 6)

##### in this format which data is ready to be used, we make appropriate changes (like adding prompt) to feed it to the data loader for later usage.

In [47]:
agent_prompts, agents, target_prompts, targets = [], [], [], []
for i in range(len(srl_df)):
    line = srl_df.iloc[i]
    agent_prompts.append(f" sentence is: {line['sentences']} find agent for this verb: {line['verbs']}")
    agents.append(line['agents'])
    target_prompts.append(f" sentence is: {line['sentences']} find target for this verb: {line['verbs']}")
    targets.append(line['targets'])
srl_df = []
srl_df = pd.DataFrame({'agent_prompt': agent_prompts, 'agent': agents, 'target_prompt': target_prompts, 'target': targets})
srl_df

agent_prompt  \
0       sentence is: Trinity said 0 it plans *PRO*-1 ...   
1       sentence is: In an Oct. 19 review of `` The M...   
2       sentence is: In an Oct. 19 review of `` The M...   
3       sentence is: In an Oct. 19 review of `` The M...   
4       sentence is: In an Oct. 19 review of `` The M...   
...                                                  ...   
63192   sentence is: `` We emphasize discipline becau...   
63193   sentence is: `` We want *PRO*-2 to see Nelson...   
63194   sentence is: `` We want *PRO*-2 to see Nelson...   
63195   sentence is: `` We want *PRO*-2 to see Nelson...   
63196   sentence is: `` We want *PRO*-2 to see Nelson...   

                                                   agent  \
0                                             it *PRO*-1   
1                                                          
2                                   Revitalized Classics   
3                                                          
4                                                          
...                                                  ...   
63192  Andrew Mlangeni, another early Umkhonto leader...   
63193                                                 We   
63194                                         We *PRO*-2   
63195                                                      
63196                                                 we   

                                           target_prompt  \
0       sentence is: Trinity said 0 it plans *PRO*-1 ...   
1       sentence is: In an Oct. 19 review of `` The M...   
2       sentence is: In an Oct. 19 review of `` The M...   
3       sentence is: In an Oct. 19 review of `` The M...   
4       sentence is: In an Oct. 19 review of `` The M...   
...                                                  ...   
63192   sentence is: `` We emphasize discipline becau...   
63193   sentence is: `` We want *PRO*-2 to see Nelson...   
63194   sentence is: `` We want *PRO*-2 to see Nelson...   
63195   sentence is: `` We want *PRO*-2 to see Nelson...   
63196   sentence is: `` We want *PRO*-2 to see Nelson...   

                                                  target  
0                                               delivery  
1                                               Classics  
2                                              the Stage  
3                                 the role of Celimene *  
4      the role of Celimene, played * by Kim Cattrall...  
...                                                  ...  
63192  We emphasize discipline because we know that t...  
63193  *PRO*-2 to see Nelson Mandela and all our comr...  
63194  Nelson Mandela and all our comrades out of prison  
63195                                             we *-1  
63196                                               them  

[63197 rows x 4 columns]

In [48]:
srl_dataset = CustomDataset(srl_df, tokenizer, max_length4text= MAX_INPUT_LENGTH, max_length4label= MAX_LABEL_LENGTH)
srl_data_loader = DataLoader(srl_dataset, batch_size= BATCH_SIZE, shuffle=True)

In [49]:
 model.train()
for epoch in range(EPOCH):
    losses = []
    for batch in srl_data_loader:

        agent_input = batch['agent_input_id']
        agent_attention_mask = batch['agent_attention_mask']

        target_input = batch['target_input_id']
        target_attention_mask = batch['target_attention_mask']

        agent_id = batch['agent_id']
        agent_mask = batch['agent_mask']

        target_id = batch['target_id']
        target_mask = batch['target_mask']

        optimizer.zero_grad()

        agent_output = model(agent_input, attention_mask=agent_attention_mask, labels=agent_id)
        target_output = model(target_input, attention_mask=target_attention_mask, labels=target_id)

        agent_loss = agent_output.loss
        target_loss = target_output.loss
        losses.append(agent_loss.item())
        losses.append(target_loss.item())

        agent_loss.backward()
        target_loss.backward()
        optimizer.step()

    print(f'loss: {np.mean(losses)}\n end of epoch{epoch}. \n')

loss: 0.32702495315937397
 end of epoch0. 

loss: 0.21392199849727553
 end of epoch1. 

loss: 0.18635298289349775
 end of epoch2. 

loss: 0.16881466145727406
 end of epoch3. 

loss: 0.15460037957067907
 end of epoch4. 

loss: 0.1447543637493413
 end of epoch5. 

loss: 0.1357652060233522
 end of epoch6. 

loss: 0.13077482092885384
 end of epoch7. 

loss: 0.12256063378974501
 end of epoch8. 

loss: 0.11930875427193152
 end of epoch9. 



In [51]:
torch.save(model, '70srl_t5.pt')

### can be skipped if the model is accessible 

In [37]:
model = torch.load('/kaggle/input/srl_traned_t5/pytorch/10-percent/1/srl_mpqa.pt')

In [36]:
# fold0 training process of normal mpqa:
train_data(train_data_loader, val_data_loader)

UnboundLocalError: local variable 'mpqa_trained_output' referenced before assignment

In [117]:
target_accuracy = evaluate_model(model, test_data_loader, 'target')
agent_accuracy = evaluate_model(model, test_data_loader, 'agent')
print(f'f1 Agent: {agent_accuracy}. f1 Target: {target_accuracy}\n\n----------------------')

matched: 455430, prediction_len:797449, actual_len:797449 

matched: 633137, prediction_len:797449, actual_len:797449 

f1 Agent: 0.7939529675251961. f1 Target: 0.5711086226203808

----------------------


fold0_agent = 0.8197088465845465
fold0_target = 0.5834266517357223

fold1_agent = 0.7872340425531915
fold1_target = 0.5599104143337066

fold2_agent = 0.7849944008958566
fold2_target = 0.568868980963046

fold3_agent = 0.7950727883538634
fold3_target = 0.5789473684210527

fold4_agent = 0.7939529675251961
fold4_target = 0.571108622620380

## agent: 79.68
## target: 57.16